pkg/server/io/io.go

```diff
@@ -263,56 +275,66 @@ func (c *ContainerIO) Attach(stdin io.Reader, stdout, stderr io.WriteCloser) err
 	stdoutKey := streamKey(c.id, "attach-"+key, Stdout)
 	stderrKey := streamKey(c.id, "attach-"+key, Stderr)
 
-	var stdinCloser io.Closer
-	if c.stdinPath != "" && stdin != nil {
+	var stdinRC io.ReadCloser
+	if c.stdinPath != "" && opts.Stdin != nil {
 		f, err := fifo.OpenFifo(c.closer.ctx, c.stdinPath, syscall.O_WRONLY|syscall.O_NONBLOCK, 0700)
 		if err != nil {
 			return err
 		}
+		// Create a wrapper of stdin which could be closed. Note that the
+		// wrapper doesn't close the actual stdin, it only stops io.Copy.
+		// The actual stdin will be closed by stream server.
+		stdinRC = cioutil.NewWrapReadCloser(opts.Stdin)
 		// Also increase wait group here, so that `closer.Wait` will
 		// also wait for this fifo to be closed.
 		c.closer.wg.Add(1)
 		wg.Add(1)
 		go func(w io.WriteCloser) {
-			if _, err := io.Copy(w, stdin); err != nil {
+			if _, err := io.Copy(w, stdinRC); err != nil && err != io.ErrClosedPipe {
 				glog.Errorf("Failed to redirect stdin for container attach %q: %v", c.id, err)
 			}
 			w.Close()
 			glog.V(2).Infof("Attach stream %q closed", stdinKey)
-			// No matter what, when stdin is closed (io.Copy unblock), close stdout and stderr
-			if stdout != nil {
-				c.stdout.Remove(stdoutKey)
-			}
-			if stderr != nil {
-				c.stderr.Remove(stderrKey)
+			if opts.StdinOnce && !opts.Tty {
+				// Close container stdin if StdinOnce is specified, and tty is false.
+				if err := opts.CloseStdin(); err != nil {
+					glog.Errorf("Failed to close stdin for container %q: %v", c.id, err)
+				}
+			} else {
+				// No matter what, when stdin is closed (io.Copy unblock), close stdout and stderr
```

Comment:
not following the logic here.. Why only close stdout and stderr if !opts.StdinOnce || opts.Tty ?

Reply:
I also don't quite the logic...
This is a weird behavior, but `kubectl` is relying on it. :(

Reply:
kk
// due to kubectl behavior requirements, when !opts.StdinOnce || opts.Tty we have to hold off closing stdout and stderr until wg.Done() (all streams are closed) this replicates current docker shim behavior.

Reply:
Will do.